File: Training/OOPFundamentals/C#/Exercises/after/Exceptions/Glass.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Exceptions
{
    public class Glass
    {
        private float _maxGallons;
        private float _currentGallons;

        public Glass(float maxGallons)
        {
            _maxGallons = maxGallons;
        }

        public void AddWater(float gallonsToAdd)
        {
            float possibleGallons = _currentGallons + gallonsToAdd;
            if (possibleGallons > _maxGallons)
            {
                throw new GlassOverflownException("glass capacity exceeded; you have spilt " + (possibleGallons - _maxGallons) + " gallons");
            }
            else
            {
                _currentGallons = possibleGallons;
            }
        }
    }
}

File: Training/OOPFundamentals/Java/Exercises/after/07 Exceptions/src/Glass.java
public class Glass {

    private float maxGallons;
    private float currentGallons;

    public Glass(float maxGallons) {
        this.maxGallons = maxGallons;
    }

    public void addWater(float gallonsToAdd) throws GlassOverflownException {
        float waterAfterFilling = this.currentGallons + gallonsToAdd;
        if(waterAfterFilling > maxGallons) {
            this.currentGallons = maxGallons;
            throw new GlassOverflownException("glass capacity exceeded;" +
                    " you have spilt " + (waterAfterFilling - this.maxGallons) + " gallons");
        } else {
            this.currentGallons = waterAfterFilling;
        }
    }

}
